### PR TITLE
Doc improvements, typo fixes

### LINKED ISSUE #19: Unclear security guarantees regarding pointer arithmetic in unsafe code
Hi,

Currently in `try_array_init`, in the code for both the "need drop" and "needs not drop" cases, there are the following comments:

https://github.com/Manishearth/array-init/blob/ffb5f8ed8d7532c15109cdfa6b899058a7ee5acc/src/lib.rs#L155-L156

https://github.com/Manishearth/array-init/blob/ffb5f8ed8d7532c15109cdfa6b899058a7ee5acc/src/lib.rs#L205-L206

which are reformulations of those, originally written by @vorner:

https://github.com/Manishearth/array-init/blob/d1279d6d4cc0b271122bd054a00d07c68c3a111c/src/lib.rs#L97-L99

https://github.com/Manishearth/array-init/blob/d1279d6d4cc0b271122bd054a00d07c68c3a111c/src/lib.rs#L138-L140

I'm not sure of what these comments mean and what are their implications in terms of safety. Especially, `pointer::add` has the following code:

```rust
    pub const unsafe fn add(self, count: usize) -> Self
    where
        T: Sized,
    {
        // SAFETY: the caller must uphold the safety contract for `offset`.
        unsafe { self.offset(count as isize) }
    }
```

which states that *the caller must uphold the safety contract for `offset`*, safety contract which [is](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.offset), if I'm correct:

> If any of the following conditions are violated, the result is Undefined Behavior:
> 
>   - Both the starting and resulting pointer must be either in bounds or one byte past the end of the same allocated object. Note that in Rust, every (stack-allocated) variable is considered a separate allocated object.
> 
>   - The computed offset, in bytes, cannot overflow an isize.
> 
>   - The offset being in bounds cannot rely on "wrapping around" the address space. That is, the infinite-precision sum, in bytes must fit in a usize.

From  my understanding, the risk would be that `Array::Item` has a size (in bytes) greater than `isize::MAX`, because we only increase the pointer by 1. Yet, the reference [states](https://doc.rust-lang.org/stable/reference/types/numeric.html#machine-dependent-integer-types) that:

> The isize type is a signed integer type with the same number of bits as the platform's pointer type. The theoretical upper bound on object and array size is the maximum isize value. This ensures that isize can be used to calculate differences between pointers into an object or array and can address every byte within an object along with one byte past the end.

Hence, I'm not sure that any overflow can happen here. That being said, if it can happen, I don't think using `add` rather than `offset` safeguards against it.

Let me know if I can help in any way

### PR DESCRIPTION
**Pull request by krtab**
Hi !

This PR is a bit of a mess, if need be I can split it up in more atomic sub PR.

Anyway, it mainly consist of adding documentation, fixing some typos, and an unfortunate `cargo fmt` at the beginning which makes for a lot of the changes.

I think there are two commits that are a bit more important than others:
 - https://github.com/Manishearth/array-init/commit/6b56e473dc0801c0c9a256c252d557f59da28c9e (fixes: #19)
 - https://github.com/Manishearth/array-init/commit/1fe42c7b3c7c09ecabcd58b45d813f4c9c60b566

More broadly speaking, I'm interested in putting in some work to bring this crate to a more 1.0 state

### CODE REVIEW

**Comment by Manishearth** [comment]
That would be great, thanks! Might take a little bit to get around to review, but thank you for the PR!

File: src/lib.rs [hunk 1]
```diff
@@ -89,7 +89,7 @@ where
     enum Unreachable {}
 
     try_array_init(
-        // monomorphise into an unfallible version
+        // monomorphise into an unfailible version
```
Comment (Manishearth) [inline src/lib.rs]: the word should be _infallible_

File: src/lib.rs [hunk 2]
```diff
@@ -66,8 +66,8 @@ pub unsafe trait IsArray {
 #[inline]
 /// Initialize an array given an initializer expression.
 ///
-/// The initializer is given the index of the element. It is allowed
-/// to mutate external state; we will always initialize the elements in order.
+/// The initializer is given the index (between 0 and `Array:len() - 1` included) of the element. It is allowed
```
Comment (Manishearth) [inline src/lib.rs]: I don't think this commit is particularly helpful. "In order" should suffice, there is no need to explain the bounds

File: src/lib.rs [hunk 3]
```diff
@@ -168,6 +168,11 @@ where
     Array: IsArray,
     F: FnMut(usize) -> Result<Array::Item, Err>,
 {
+    // The implementation differentiate two cases:
```
Comment (Manishearth) [inline src/lib.rs]:
```suggestion
    // The implementation differentiates two cases:
```

**Review by Manishearth — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (krtab) [reply src/lib.rs]: My bad, I double checked, but I should have triple checked.

File: src/lib.rs (hunk 2 above)
Comment (krtab) [reply src/lib.rs]: Agreed

**Comment by krtab** [comment]
Hi,

Thanks for you review, I have modified files accordingly. I guess a rebase would be cleaner, I'll open an other PR. Just let me know if you want me to cargo fmt the whole file or not;

File: src/lib.rs [hunk 4]
```diff
@@ -89,7 +89,7 @@ where
     enum Unreachable {}
 
     try_array_init(
-        // monomorphise into an unfailible version
+        // monomorphise into an unfallible version
```
Comment (Manishearth) [inline src/lib.rs]:
```suggestion
        // monomorphise into an infallible version
```

No, you were correct in that "unfallible" is not a word, your fix was incorrect: it should be "infallible"

**Pull request merged**